crypto/attachment.go

```diff
@@ -13,6 +15,178 @@ import (
 	"github.com/pkg/errors"
 )
 
+// AttachmentProcessor2 keeps track of the progress of encrypting an attachment
+// (optimized for encrypting large files).
+// With this processor, the caller has to first allocate
+// a buffer large enough to hold the whole data packet.
+type AttachmentProcessor2 struct {
+	keyPacket        []byte
+	dataLength       int
+	plaintextWriter  io.WriteCloser
+	ciphertextWriter *io.PipeWriter
+	err              error
+	done             sync.WaitGroup
+}
+
+// GetKeyPacket returns the key packet for the attachment.
+// This should be called only after Finish() has been called.
+func (ap *AttachmentProcessor2) GetKeyPacket() []byte {
+	return ap.keyPacket
+}
+
+// GetDataLength returns the number of bytes in the DataPacket.
+// This should be called only after Finish() has been called.
+func (ap *AttachmentProcessor2) GetDataLength() int {
+	return ap.dataLength
+}
+
+// Process writes attachment data to be encrypted.
+func (ap *AttachmentProcessor2) Process(plainData []byte) error {
+	defer runtime.GC()
+	_, err := ap.plaintextWriter.Write(plainData)
+	return errors.Wrap(err, "gopenpgp: couldn't write attachment data")
+}
+
+// Finish tells the processor to finalize encryption.
+func (ap *AttachmentProcessor2) Finish() error {
+	defer runtime.GC()
+	if ap.err != nil {
+		return ap.err
+	}
+	if err := ap.plaintextWriter.Close(); err != nil {
+		return errors.Wrap(err, "gopengpp: unable to close plaintext writer")
+	}
+	if err := ap.ciphertextWriter.Close(); err != nil {
+		return errors.Wrap(err, "gopengpp: unable to close the dataPacket writer")
+	}
+	ap.done.Wait()
+	if ap.err != nil {
+		fmt.Printf("Error while finishing %v\n", ap.err)
+		return ap.err
+	}
+	return nil
+}
+
+// NewLowMemoryAttachmentProcessor2 creates an AttachmentProcessor which can be used
+// to encrypt a file. It takes an estimatedSize and filename as hints about the
+// file and a buffer to hold the DataPacket.
+// It is optimized for low-memory environments and collects garbage every megabyte.
+// Make sure that the dataBuffer is large enough to hold the whole data packet
+// otherwise Finish() will return an error.
+func (keyRing *KeyRing) NewLowMemoryAttachmentProcessor2(
+	estimatedSize int, filename string, dataBuffer []byte,
+) (*AttachmentProcessor2, error) {
+	if dataBuffer == nil || cap(dataBuffer) == 0 {
+		return nil, errors.New("gopenpgp: can't give a nil or empty buffer to process the attachement")
+	}
+
+	// forces the gc to be called often
+	debug.SetGCPercent(10)
+
+	attachmentProc := &AttachmentProcessor2{}
+
+	// hints for the encrypted file
+	isBinary := true
+	modTime := uint32(GetUnixTime())
```

Comment:
Why uint32 to then cast to an int64?

Reply:
I don't know, I just copied the way it's done in the old processor

Reply:
I suspected, probably we should change this. @twiss is there any reason I'm missing here?

Reply:
Well, timestamps in OpenPGP are indeed uint32 (:sweat_smile:).
The casting from a time to a number and back seems unnecessary indeed though, this can be left to go-crypto.

Reply:
Ok I'll remove the unnecessary casting then